src/engine/sync/midimaster.cpp

```diff
@@ -0,0 +1,131 @@
+#include <QtDebug>
+
+#include "engine/sync/midimaster.h"
+
+#include "engine/sync/enginesync.h"
+#include "controllers/midi/midisourceclock.h"
+#include "controllinpotmeter.h"
+#include "controlobject.h"
+#include "controlpushbutton.h"
+#include "configobject.h"
+#include "util/time.h"
+
+MidiMasterClock::MidiMasterClock(const char* pGroup, SyncableListener* pEngineSync)
+        : m_group(pGroup),
+          m_pEngineSync(pEngineSync),
+          m_mode(SYNC_NONE) {
+    m_pMidiSourceClockBpm.reset(new ControlObject(ConfigKey(pGroup, "bpm")));
+    m_pMidiSourceClockLastBeatTime.reset(
+            new ControlObject(ConfigKey(pGroup, "last_beat_time")));
+    m_pMidiSourceClockBeatDistance.reset(
+            new ControlObject(ConfigKey(pGroup, "beat_distance")));
+
+    m_pMidiSourceClockRunning.reset(
+            new ControlPushButton(ConfigKey(pGroup, "run")));
+
+    m_pSyncMasterEnabled.reset(
+            new ControlPushButton(ConfigKey(pGroup, "sync_master")));
+    m_pSyncMasterEnabled->setButtonMode(ControlPushButton::TOGGLE);
+    m_pSyncMasterEnabled->connectValueChangeRequest(
+            this, SLOT(slotSyncMasterEnabledChangeRequest(double)),
+            Qt::DirectConnection);
+
+    m_pMidiSourceClockSyncAdjust.reset(
+            new ControlLinPotmeter(ConfigKey(pGroup, "sync_adjust"), -.5, .5,
+                                   0.1, 0.01, /*allow oob*/ true));
+}
+
+MidiMasterClock::~MidiMasterClock() {
+};
+
+void MidiMasterClock::notifySyncModeChanged(SyncMode mode) {
+    // Syncable has absolutely no say in the matter. This is what EngineSync
+    // requires. Bypass confirmation by using setAndConfirm.
+    m_mode = mode;
+    m_pSyncMasterEnabled->setAndConfirm(mode == SYNC_MASTER);
+}
+
+void MidiMasterClock::notifyOnlyPlayingSyncable() {
+    // No action necessary.
+}
+
+void MidiMasterClock::requestSyncPhase() {
+    // TODO: maybe tell MidiSourceClock to reset which tick is the beat tick??
+    // but really, it's a read-only clock.
+}
+
+void MidiMasterClock::slotSyncMasterEnabledChangeRequest(double state) {
+    bool currentlyMaster = getSyncMode() == SYNC_MASTER;
+
+    if (state > 0.0) {
+        if (currentlyMaster) {
+            // Already master.
+            return;
+        }
+        m_pEngineSync->requestSyncMode(this, SYNC_MASTER);
+    } else {
+        // TODO: midi follower (clock out?)
+        m_pEngineSync->requestSyncMode(this, SYNC_NONE);
+    }
+}
+
+double MidiMasterClock::getBeatDistance() const {
+    qint64 last_beat = static_cast<qint64>(m_pMidiSourceClockLastBeatTime->get());
+    double raw_percent = MidiSourceClock::beatFraction(last_beat, Time::elapsed(),
+                                                       m_pMidiSourceClockBpm->get());
+    raw_percent += m_pMidiSourceClockSyncAdjust->get();
```

Comment:
I tried moving the sync adjustment into the sourceclock, but that caused a lot of problems.  Tthe clock class doesn't care about sync, it is just trying to come up with accurate bpm numbers.   Trying to adjust timestamps inside midisourceclock is dangerous because the user may call time->now() and all the calculations will be off because the stamps inside midisoureclock have an offset applied.   Also if the user needs to adjust the sync then all ringbuffer data inside midisourceclock will be incorrect and the reported bpm numbers will be wildly incorrect.

It's better to have midisourceclock be pure and then any sync adjustment be done inside the midi sync master.